BCI tests for python with trainer modules (#108)
SUSE/BCI-tests: tests/trainers/README.md

BCI tests modules
=================

> Description

Reviewer: I don't think that this fits into the README. However, this is a pretty good guide and it would be a shame not to use it. Can I include this part in the documentation of [pytest_container](https://github.com/dcermak/pytest_container/)?

Author: Sure, thank you!

Author: I moved the copy syntax description in a DOCUMENT separated from the README file containing the only test specifications. I used the md markdown formatting.

    ...
    """

    CONTAINER1 = pytest.param(
        DerivedContainer(
            base=container_from_pytest_param(CONTAINER),
            containerfile=DOCKERFIL,
        ),
        marks=CONTAINER.marks,
    )

    CONTAINER_IMAGES_NEW = [ CONTAINER1, ... ]

Then we parametrize the new container for each test and assign it to the auto_container_per_test fixtures, adding this last as input parameter of the test;

:..

    @pytest.mark.parametrize("auto_container_per_test", [CONTAINER_IMAGES_NEW], indirect=True)
    def test_ testname(auto_container_per_test, ...):
        ...

In the test code we can run commands in the new container using the testinfra commands (See <https://testinfra.readthedocs.io/en/latest/modules.html>).

I.e.:

:..

    auto_container_per_test.connection.run_expect([0], "python3 filename.py")

2. Copy local files into the running container, using the 'podman|docker' cp command, via testinfra:

:..

    def test_<nnn>(auto_container_per_test, host, container_runtime):

        CID = auto_container_per_test.container_id

        host.run_expect( [0],
            f"{container_runtime.runner_binary} cp {DIR}/{FILE}  {CID}:{CDIR}")

> Tests for BCI python

Reviewer: This is a really great description!

Author: Thank you! 
Indeed in this description the _auto_container_per_test_ parameter could also be simply _container_.

# test-for-py

> tests for python

Reviewer: You can omit this file, this really adds no additional information.

Author: I was going to update the file with some description of those new modules

Reviewer: Ah, thanks, that would be much better of course.

Author: README file updated

BCI tests modules
=================

Description
-----------

> This directory contains modules and files that shall be copied into the container under test, for the execution of specific exercises and examples during the BCI validation .

Reviewer: ```suggestion
This directory contains modules and files that shall be copied into the container under test, for the execution of specific exercises and examples during the BCI validation.
```

> See the pytest_container/source/usage.rst documentation about the _copy to containers_.

Reviewer: ```suggestion
For information how to copy files into containers, consult the [pytest_container documentation](https://github.com/dcermak/pytest_container/blob/main/source/usage.rst#copying-files-into-containers=)
```

Tests for BCI python
--------------------

1. Communication examples: 

Module to check communication capabilities of python functionalities. 

The *get_file_www* function test that the python _wget_ [library](https://pypi.org/project/wget/) is able to fetch files from a webserver.

We use the _download_ method to get a specific file from a remote url.

Input parameters of this function are: (1) "URL/FIL" , (2) "DIR", where:

- URL : the http remote url
- FIL : the remote file to get
- DIR : the directory in the container receiving FIL 

> Expected for this test: FIL present in DIR.

Reviewer: All of this belongs into the docstring of `get_file_www` as then sphinx would be able to pick them up.

Author: Done

1. Tensorflow examples:

This is a pyhton training module making use of tensorflow library and some related tutorials, under Apache License.

The *tensorflow_example_1* function is derived from the code in the 'TensorFlow 2 - quickstart for beginners' [tutorial](https://github.com/tensorflow/docs/blob/master/site/en/tutorials/quickstart/beginner.ipynb)

This function, using the Tensorflow library, trains a machine learning model and evaluates accuracy:

- Load a ML model using a prebuilt dataset with the Keras API
- Build a neural network ML model that classifies images
- Trains this neural network
- Evaluates the accuracy of the model

> Expected for this test: accuracy greather than 0.9 and loss lower than 0.1.

Reviewer: I'd suggest to move this into the docstring of the `tensorflow_example_1` function.

Author: Do you mean to move all the function description from README file to the related code module itself ? 
In this case the README file will be reduced to a mention referring to the docstring. For me is ok.

Reviewer: I'd just mention in the readme that these test files are copied into the container under test and that's it.

Author: Done
